Sources/CacheAdvance/CacheAdvance.swift

```diff
@@ -73,9 +73,7 @@ public final class CacheAdvance<T: Codable> {
     /// - Parameter message: A message to write to disk. Must be smaller than both `maximumBytes - FileHeader.expectedEndOfHeaderInFile` and `MessageSpan.max`.
     public func append(message: T) throws {
         try setUpFileHandlesIfNecessary()
-        guard try header.canOpenFile() else {
-            throw CacheAdvanceError.incompatibleHeader
-        }
+        try header.openFile()
```

Comment:
@bachand the changes in this file are the ones that cause the header-reading errors to be propagated. Previously we'd swallow whatever errors were encountered when calling `header.canOpenFile`. Now we propagate those errors directly.

Reply:
@dfed I feel silly posting this I feel like I must be missing something. I guess I'm not seeing what error we are swallowing. If `header.canOpenFile()` throws we propagate that error upward since the `guard` only applies if `header.canOpenFile()` returns normally.

<img width="1072" alt="Screen Shot 2021-01-11 at 8 42 09 PM" src="https://user-images.githubusercontent.com/1791049/104270560-a6a55480-544d-11eb-8a1d-ead4c4c2a3ba.png">

So I am pretty sure that if `memoizedMetadata()` (called within `canOpenFile()`) were to throw we would propagate that error outwards to the client.

What am I missing? 😖

Reply:
Good writeup @bachand! So... I think I was blinded by the fact because I'm seeing reports of `CacheAdvanceError.incompatibleHeader` being thrown and so I assumed we must be squashing a better error. If we're properly propagating errors encountered... then I have absolutely no clue how we're hitting this condition.

I'm open to reverting this part of the change, since you're right it doesn't make a functional difference. I do like the idea of `CacheHeaderHandle` throwing the error since it's the code doing the comparison, but you're right that doing so isn't a requirement.

Now I'm really curious what the version we're encountering is. If I had to guess: the versions will be somewhat random, and we're encountering files that shouldn't be where we're looking.

Reply:
Can you help me understand how the versions could be random? I'm not following that part.

Reply:
The following code will throw a `CacheAdvanceError.incompatibleHeader(persistedVersion: 116)`:
```
FileManager.default.createFile(
    atPath: testFileLocation.path,
    contents: "this is a somewhat long string that is not real CacheAdvance data.".data(using: .utf8)!,
    attributes: nil)
try CacheHeaderHandle(
    forReadingFrom: testFileLocation,
    maximumBytes: 5000,
    overwritesOldMessages: false).checkFile()
```

To determine the header's version, we read the first `UInt8.storageLength` of whatever file a `CacheAdvance` instance is pointed at and turn that value into an integer. If we're reading a file that wasn't written by `CacheAdvance`, then the value of the version in the file "header" is effectively random/meaningless.

Reply:
Ah got it. We may want to consider adding a unique (or as-unique-as-possible) [signature](https://en.wikipedia.org/wiki/List_of_file_signatures) value to the start of the file. That could mitigate this problem significantly. If the file doesn't have the signature value we can bail out immediately and not try to parse it.

Reply:
Yup! Thankfully we reserved 38bytes in our header, so we have room to encode a signature without making a breaking change. Though if we take this approach we'd need to enable an upgrade path for files written using the version without this info. Food for future thought.

In the meantime, the version matching is effectively a signature since we know all the possible valid values. It's a short and not-very-complex signature, but it's what we've got thus far 😅

Reply:
Interesting finding: the file size of the file with an incompatible header is _always_ 64bytes. Meaning that it's a file to which we never wrote any messages.

I still don't understand why the header file has an incompatible header (I screwed up and forgot to log the header version). But the good news is we're never losing messages.